go/vt/vtgate/planbuilder/select2.go

```diff
@@ -0,0 +1,367 @@
+	}
+	isLeft := false
+	if join.Join == sqlparser.LeftJoinStr {
+		isLeft = true
+	}
+	assignOrder(rplanBuilder, lplanBuilder.Order())
+	return &JoinBuilder{
+		IsLeft: isLeft,
+		order:  rplanBuilder.Order(),
+		Left:   lplanBuilder,
+		Right:  rplanBuilder,
+	}, lsymbols, nil
+}
+
+// assignOrder sets the order for the nodes of the tree based on the
+// starting order.
+func assignOrder(planBuilder PlanBuilder, order int) {
+	switch planBuilder := planBuilder.(type) {
+	case *JoinBuilder:
+		assignOrder(planBuilder.Left, order)
+		assignOrder(planBuilder.Right, planBuilder.Left.Order())
+		planBuilder.order = planBuilder.Right.Order()
+	case *RouteBuilder:
+		planBuilder.order = order + 1
+	}
+}
+
+// joinRoutes attempts to join two RouteBuilder objects into one.
+// If it's possible, it produces a joined RouteBuilder.
+// Otherwise, it's a JoinBuilder.
+func joinRoutes(lRouteBuilder *RouteBuilder, lsymbols *SymbolTable, rRouteBuilder *RouteBuilder, rsymbols *SymbolTable, join *sqlparser.JoinTableExpr) (PlanBuilder, *SymbolTable, error) {
+	if lRouteBuilder.Route.Keyspace.Name != rRouteBuilder.Route.Keyspace.Name {
+		return makeJoinBuilder(lRouteBuilder, lsymbols, rRouteBuilder, rsymbols, join)
+	}
+	if lRouteBuilder.Route.PlanID == SelectUnsharded {
+		if rRouteBuilder.Route.PlanID == SelectUnsharded {
+			// Two Routes from the same unsharded keyspace can be merged.
+			return mergeRoutes(lRouteBuilder, lsymbols, rsymbols, join)
+		}
+		return makeJoinBuilder(lRouteBuilder, lsymbols, rRouteBuilder, rsymbols, join)
+	}
+	// lRouteBuilder is a sharded route. It can't merge with an unsharded route.
+	if rRouteBuilder.Route.PlanID == SelectUnsharded {
+		return makeJoinBuilder(lRouteBuilder, lsymbols, rRouteBuilder, rsymbols, join)
+	}
+	// TODO(sougou): Handle special case for SelectEqual and SelectKeyrange.
+	// Both RouteBuilder are sharded routes. Analyze join condition for merging.
+	return joinShardedRoutes(lRouteBuilder, lsymbols, rRouteBuilder, rsymbols, join)
+}
+
+// mergeRoutes makes a new RouteBuilder by joining the left and right
+// nodes of a join. This is called if two routes can be merged.
+func mergeRoutes(lRouteBuilder *RouteBuilder, lsymbols, rsymbols *SymbolTable, join *sqlparser.JoinTableExpr) (PlanBuilder, *SymbolTable, error) {
+	lRouteBuilder.From = join
+	err := lsymbols.Merge(rsymbols, lRouteBuilder)
+	if err != nil {
+		return nil, nil, err
+	}
+	return lRouteBuilder, lsymbols, nil
+}
+
+// joinShardedRoutes tries to join two sharded routes into a RouteBuilder.
+// If a merge is possible, it builds one using lRouteBuilder as the base route.
+// If not, it builds a JoinBuilder instead.
+func joinShardedRoutes(lRouteBuilder *RouteBuilder, lsymbols *SymbolTable, rRouteBuilder *RouteBuilder, rsymbols *SymbolTable, join *sqlparser.JoinTableExpr) (PlanBuilder, *SymbolTable, error) {
+	onFilters := appendFilters(nil, join.On)
+	for _, filter := range onFilters {
+		if !isSameRoute(filter, lsymbols, rsymbols) {
```

Comment:
I find the logic easier to read if the inside of the loop was written like:
If (isSameRoute(...) {
   return mergeRoutes(...)
}